proxy_ssh.go

```diff
@@ -0,0 +1,287 @@
+		s.logger.Error(errors.Wrap(err, "ssh.prxy"))
+		return (err)
+	}
+
+	go ssh.DiscardRequests(reqs)
+
+	for ch := range chans {
+
+		sshClientChan, clientReq, err := clientConn.OpenChannel(ch.ChannelType(), ch.ExtraData())
+		if err != nil {
+			s.logger.Error(errors.Wrap(interpreter(" Could not accept client channel: ", err), "ssh.prxy"))
+			return err
+		}
+
+		sshServerChan, serverReq, err := ch.Accept()
+		if err != nil {
+			s.logger.Error(errors.Wrap(interpreter(" Could not accept server channel: ", err), "ssh.prxy"))
+			return err
+		}
+
+		// Connect requests of ssh server and client
+		go func() {
+			s.logger.Debug("[prxy.ssh] Waiting for request")
+
+		r:
+			for {
+				var req *ssh.Request
+				var dst ssh.Channel
+
+				select {
+				case req = <-serverReq:
+					dst = sshClientChan
+				case req = <-clientReq:
+					dst = sshServerChan
+				}
+
+				// Check if connection is closed
+				if req == nil {
+					s.logger.Debug("[prxy.ssh] SSH Request is nil")
+					return
+				}
+
+				s.logger.Debugf("[prxy.ssh] Request: \n\n%s %s %s %s\n\n", dst, req.Type, req.WantReply, req.Payload)
+				b, err := dst.SendRequest(req.Type, req.WantReply, req.Payload)
+				if err != nil {
+					s.logger.Error(errors.Wrap(err, "ssh.prxy"))
+				}
+
+				if req.WantReply {
+					req.Reply(b, nil)
+				}
+
+				switch req.Type {
+				case "exit-status":
+					break r
+				case "exec":
+					s.logger.Debug("[prxy.ssh] SSH request 'EXEC' is not supported")
+				default:
+					s.logger.Debugf("[prxy.ssh] %s", req.Type)
+				}
+			}
+
+			sshServerChan.Close()
+			sshClientChan.Close()
+		}()
+
+		var wrappedServerChan io.ReadCloser = sshServerChan
+		var wrappedClientChan io.ReadCloser = sshClientChan
+
+		defer wrappedServerChan.Close()
+		defer wrappedClientChan.Close()
+
+		if s.wrapFn != nil {
+			wrappedClientChan, err = s.wrapFn(serverConn, sshClientChan)
+		}
+
+		go io.Copy(sshClientChan, wrappedServerChan)
+		go io.Copy(sshServerChan, wrappedClientChan)
+	}
+
+	if s.closeFn != nil {
+		s.closeFn(serverConn)
+	}
+
+	return nil
+}
+
+// TODO: Use of existing key
+func (s *SSHProxy) SSHKeyGen() ([]byte, error) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2014)
+	if err != nil {
+		s.logger.Error(errors.Wrap(err, "ssh.prxy"))
+		return nil, err
+	}
+	err = priv.Validate()
+	if err != nil {
+		s.logger.Error(errors.Wrap(interpreter("Validation failed.", err), "ssh.prxy"))
+		return nil, err
+	}
+
+	priv_der := x509.MarshalPKCS1PrivateKey(priv)
+
+	priv_blk := pem.Block{
+		Type:    "RSA PRIVATE KEY",
+		Headers: nil,
+		Bytes:   priv_der,
+	}
+
+	RSA_Key := pem.EncodeToMemory(&priv_blk)
+
+	// Shot to validating private bytes
+	_, err = ssh.ParsePrivateKey(RSA_Key)
+	if err != nil {
+		s.logger.Error(errors.Wrap(err, "ssh.prxy"))
+		return nil, err
+	}
+	return RSA_Key, nil
+}
+
+func interpreter(msg string, err error) error {
+	return errors.New(fmt.Sprintf("%s  %s\n", msg, err))
+}
+
+func (rs *ReadSession) Read(p []byte) (n int, err error) {
```

Comment:
Yes same package, another thing I need to wrap a ssh channel to custom type with io.Reader, so that I could look into io.Copy. : https://github.com/mushorg/glutton/pull/51/files#diff-69a47fd69d5b7a436ed294d80b5228fcR107